addons/website_sale/models/sale_order.py

```diff
@@ -409,3 +418,38 @@ def _get_shop_warning(self, clear=True):
         if clear:
             self.shop_warning = ''
         return warn
+
+    def _filter_can_send_abandoned_cart_mail(self, website):
+        abandoned_datetime = datetime.utcnow() - relativedelta(hours=website.cart_abandoned_delay)
+
+        sales_done_after_self_create_date = self.env['sale.order'].search([
+            ('state', '=', 'sale'),
+            ('partner_id', 'in', self.partner_id.ids),
+            ('create_date', '>=', abandoned_datetime),
+            ('website_id', '=', website.id),
+        ])
+        related_sales_per_partner = defaultdict(lambda: self.env['sale.order'])
+        for sale in sales_done_after_self_create_date:
+            related_sales_per_partner[sale.partner_id] |= sale
+
+        # Customer needs to be signed in otherwise the mail address is not known.
+        # We therefore consider only sales with a known mail address.
+
+        # If user has manually sent a recovery email, the mail will not be sent a second time
+
+        # If a payment processing error occurred when the customer tried to complete their checkout,
+        # then the email won't be sent.
+
+        # If all the products in the checkout are free, and the customer does not visit the shipping page to add a
+        # shipping fee or the shipping fee is also free, then the email won't be sent.
+
+        # If a potential customer creates one or more abandoned sale order and then completes a sale order before
+        # the recovery email gets sent, then the email won't be sent.
+
+        return self.filtered(lambda abandoned_sale_order:
+                             abandoned_sale_order.partner_id.email
+                             and not abandoned_sale_order.cart_recovery_email_sent
+                             and not any(transaction.state == 'error' for transaction in abandoned_sale_order.transaction_ids)
+                             and abandoned_sale_order.order_line.filtered(lambda line: line.price_unit != 0.0)
```

Comment:
Slightly faster + use float_is_zero
```suggestion
                             and any(not float_is_zero(line.price_unit, precision_rounding=line.currency_id.rounding) for line in abandoned_sale_order.order_line)
```